### ISSUE TITLE
Device tree scripts for embedded devices.

### ISSUE THREAD

**Issue by necrose99** [dracutdevs/dracut#1375]
**Describe the enhancement**
A clear and concise description of what the enhancement is that you would like to see.

Script/features  to add device tree blobs to dracuts capabilities 
(Or add enhancement scripts to extend dracut,  with a simple api. For doing embedded stuff , user-friendly etc. )

could load device trees much like existing kernel modules
Ie. (Simular Treatment akin to kernel modules or ease of use) 

Example: Dracaut add rtc_ds3231.. rtc_ds1307
Adding Arm64 Real Time clock rtc_ds3231,   rtc_ds3231 into  rpi4.intramfs..   

which if using the new uefi firmware could be of use. 
And bit more dynamic than the static txt file. 

Or latte panda etc, x86, arm64,riscv , etc single board pcs. 

Anyway,  making a custom kernel with blobs and easy load dtd's into intramfs via dracut would be a nice to have.. 

Treatment of them much like nvidia or well known modules , would also make for ease of management on embedded systems.  Ie rpi4-rtc , rpi4-cmc , and carrying boards. 
Thier are several but search find probe etc esp. With a FREINDLY-ID ,  easy for users with lookup tables.  
LCD, touchscreen , camera.. 

... probe for varrious rtc hardware on product detection add correct to list for dracut to insert  into intramfs 
Load unload device trees automatically,  probe device null , unload 
Load device found keep. 
Also a plus. 

Hardware makers could script dracut to shim thier cool additions ie RasPad 3 into any kernel via dracut. Etc.
Load device trees, kernel modules for hardware no problemo.. 

Arduino fans can add the new gizmos etc..  simple install via dracut script reboot , and pain free..

Recompile newer kernels,  without breaking stuff... 
(Hopefully) 

Dracut to manage device tree blobs for embedded devices,  been hopeful feature for sometime. 
Ie. 
A Multiple arm64 device kernel.  One could switch to just the bloobs they need.. however same os image on multiple devices..  rpi4 , pinebook pro etc , (uefi and newer onboard firmware)
or add new hardware , camera etc painlessly.

**Comment by necrose99** [comment]
RockPro64/pinebook-pro gentoo , one can run uefi with sbi one newer firmware with no boot blobs ie uboot .. on sbi-ep/rom looks for uefi firmware ie grub-aarch64.efi or uboot kernel settings for chainload... 
https://github.com/pftf/RPi4 , grub2 debian or etc .. 
https://stikonas.eu/wordpress/2019/09/15/blobless-boot-with-rockpro64/

ebay happens to have bare bonse 40 core ... thunder-x2 boards/towers esp since the 80 x 2 successors Ampere-altra 
server boards also a easy to likely implement for dtd dracut initrd extensions 

stock debian can switch to grub-efi easily also.. 

_Full Uefi , able boot firmwares_
**would be the lower hanging fruit ..** 
ie lcd , camera or touch screen ... etc... 

anyboards that can boot uefi grub , **would be the easier targets** , oems could extend to fit this own sbc's if more fininiky 

as an enthusiast user I know enough to be ''Dangerous''  however most SBC's are a pain w/o uefi to update kernels and the overlay bits easily... , 
add _Dyslexia_ ,  chances of bricking something .... goes up 
let alone novice users...  or users with low english skills ie non-native speakers... 

i'm aware of some of the arm related boot joys.. less users screw with that end ... looks for grub2 or etc all the better 
'>>> grub2+uefi-enable uboot+dracut-initrd +modern kernels to boot less of a jenga tower to fu((*&^...up... ' 
--ie brick the os-image--

OEMS with fickle boots can add extensions to this as a set of plug-ins / extensibility for devs is definitely in mind...
Less hell for us users to update kernels etc.. 

OEMS with ''DKMS '' Like daughterboard overlay blobs make dracut-embedded-extensions  wala ..  item can single build dracut would shim into initramfs any add a script at boot thus uboot and crap users might break is less an issue ..   -- easy rpms/debs/etc of overlay blobs for addon-cards etc. .. 


**For now its a hopeful concept** wishlisted for sure. 
that might prove useful in easing some of the woes..  won't fix them all 
however perhaps automation of the boot blobs into initramfs were supportable might make administration for users/devs/oems less hassle ...
if you build it will they (come? ) (the oem's or other devs keep extending it to likewise ease their woes)

**Comment by devkontrol** [comment]
Can you please re-file at https://github.com/dracut-ng/dracut/issues if still relevant ?

**Comment by necrose99** [comment]
https://github.com/NumberOneGit/rpi5-uefi as initial grub2 loader or uki loader etc... 

If your not using uboot but edk2  uefi firmware loader , dracut could potentially script load devicetrees...  

(config.txt) is just loading edk rpi5-uefi rpi4 efi , others etc...  so you need other methods like grub-mkconfig  etc wich if some distros have custom portions potentially flakey if disks  stutters sdcards potentially... 
 
(config.txt) from uboot will load devicetrees 

Having dracut ram disk it means devicetree/kernel modules just load and done ... or as a uki ... and done ... 

efi boot partition 
@boot @root etc  btrfs  or etc... ext4 others ... 

Anyway it might be wiki able for embedded devices... 
 Might be a posible community contributor scema...  or wiki of examples...  

Ensure if you have efi , some blobs become part of efi boot others like rpi4 rtc add on chips 
Hats etc... 
Or riscv64...  

you can ensure  or etc devicetree objectd 
is in intrd ramdisk similar to ko modules so even on a /boot/efi , /boot  /home /[root] etc on embedded systems you could script its load... 

but it seams you need scripts as add-ons to dracut 

But loader via intrd if nvme stalls no problem,  Emmc stalls likewise,  Microsoft card...  intrd ramdisk either via grub2 or sysfs/config fs cat > devicetree$name.dtb/dtbo on boot... into kernel memory etc... 

Ive learned a few tricks since I made the suggestion... 

/usr/lib/dracut/modules.d/
#base
./embedded/$devicename/  rpi5,  rpi4 
.embedded/rpi-hats/add-on$name
# ie 5g Cellular modem hats or something...  
Tpm2.x hat over gpio... 
Rpi4_rtc.sh cp ram disk/initrd load
add_drivers+=" sdhci_bcm2835 pcie_brcmstb xhci_pci raspberrypi_hwmon bcm2835_dma "

98rtc-overlay/load-rtc-dtbo.sh. 
 #!/bin/sh
# Dracut hook: Loads RTC Device Tree Overlay via configfs

CONFIGFS_DIR="/sys/kernel/config"
DT_OVERLAY_DIR="${CONFIGFS_DIR}/device-tree/overlays/rtc"
DTBO_FILE="/etc/dtbo/rtc.dtbo"

# Ensure configfs is mounted inside the initramfs
if [ ! -d "${CONFIGFS_DIR}/device-tree" ]; then
    mkdir -p "${CONFIGFS_DIR}"
    mount -t configfs none "${CONFIGFS_DIR}" 2>/dev/null
fi

# Apply the overlay via cat if configfs interface exists
if [ -d "${CONFIGFS_DIR}/device-tree/overlays" ] && [ -f "$DTBO_FILE" ]; then
    mkdir -p "$DT_OVERLAY_DIR"
    cat "$DTBO_FILE" > "${DT_OVERLAY_DIR}/dtbo"
    
    # Optional: Verify status
    if [ "$(cat ${DT_OVERLAY_DIR}/status)" = "applied" ]; then
        echo "RTC Device Tree Overlay successfully applied via configfs."
    else
        echo "Warning: Failed to apply RTC Device Tree Overlay."
    fi
else
    echo "Error: configfs device-tree/overlays interface or DTBO file missing."
fi

 




# Force dracut to build a Unified UEFI executable
uefi="yes"

# Path to your RPi5 Device Tree Blob
uefi_fdt="/boot/dtbs/broadcom/bcm2712-rpi-5-b.dtb"


# 2. Tell GRUB 2 to load the specific Device Tree Blob (.dtb)
    devicetree /boot/dtbs/custom-device-tree.dtb

sudo mkdir -p /usr/lib/dracut/modules.d/99copy-dtb
  module-setup.sh


#!/bin/bash

check() {
    return 0 # Always include this module
}

depends() {
    return 0
}

install() {
    # Specify the source DTB path
    local DTB_SRC="/boot/dtbs/broadcom/bcm2712-rpi-5-b.dtb"
    # Specify your target location inside the physical /boot partition
    local DTB_DST="/boot/bcm2712-rpi-5-b.dtb"

    if [ -f "$DTB_SRC" ]; then
        # Copy to the actual physical boot partition during the build process
        cp "$DTB_SRC" "$DTB_DST"
    fi
}


sudo chmod +x /usr/lib/dracut/modules.d/99copy-dtb/module-setup.sh


menuentry "Linux EFI with Automated DTB" {
    insmod part_gpt
    insmod fat
    set root=(hd0,gpt1)

    linux /Image root=/dev/mmcblk0p2 rw rootwait
    devicetree /bcm2712-rpi-5-b.dtb
    initrd /initramfs.img
}


Else /etc/dracut.conf.d/uki-dtb.conf

# Enable generation of a unified UEFI executable binary
uefi="yes"

# Specify the path to your device tree blob to embed inside the EFI binary
uefi_dtb="/boot/dtbs/broadcom/bcm2712-rpi-5-b.dtb"




Ie /usr/lib/dracut/modules.d/embeded/rpi5-uefi/
99copy-dtb/module-setup.sh 

/usr/lib/dracut/modules.d/embeded/riscv64/sifive/cp-dt.ramdisk.sh 

Etc if an embedded contribution scrips by community etc... 

/etc/dracut.conf.d/sifive.conf 
#!/bin/sh
# This script runs inside the initrd ramdisk environment before the root switch

if [ -f /boot/dtbs/sifive-custom.dtb ]; then
    echo "SiFive: Copying device tree to staging area..."
    # Example: Copying to a runtime location inside the ramdisk
    cp /boot/dtbs/sifive-custom.dtb /sys/firmware/fdt 2>/dev/null || true
fi

install_items+=" /etc/dracut.conf.d/99-sifive-load.sh "
 

# View what device tree is currently loaded in memory by the hardware
cat /sys/firmware/fdt > /tmp/current_tree.dtb

# Mount a storage drive manually to copy an updated DTB over
mount /dev/nvme0n1p1 /mnt
cp /mnt/new-sifive.dtb /tmp/


#!/bin/bash
# /usr/lib/dracut/modules.d/99embedded-hardware/module-setup.sh

check() {
    return 0 # Always include for robust embedded deployments
}

depends() {
    return 0
}

install() {
    # Include absolute baseline tools needed for GPIO, hardware checking, and mapping
    inst_multiple mkdir cp find chmod tr ln grep
    
    # Optional tools for cellular HATs or debugging if they exist on host
    for cmd in gpioset gpioget dtc lshw; do
        if command -v "$cmd" >/dev/null; then inst_multiple "$cmd"; fi
    done

    # Create internal payload structure inside the initrd rootfs
    inst_dir /embedded
    
    # Copy all platform configuration subdirectories into the initrd
    cp -r "$moddir/src/"* "${initdir}/embedded/"

    # Inject the master discovery script into the early pre-mount boot hook
    inst_hook pre-mount 10 "main-discovery.sh" "$moddir/src/main-discovery.sh"
}


#!/sh
# /usr/lib/dracut/modules.d/99embedded-hardware/src/main-discovery.sh
# Runs early in initrd memory space

echo "EMBEDDED-HW: Starting multi-platform hardware discovery..." > /dev/kmsg

# 1. Read system compatibility directly from live firmware
if [ -f /sys/firmware/devicetree/base/compatible ]; then
    COMPAT=$(tr '\0' ' ' < /sys/firmware/devicetree/base/compatible)
else
    COMPAT="unknown"
fi

# 2. Match platform and branch off to specific peripheral scripts
case "$COMPAT" in
    *bcm2712*|*raspberrypi,5*)
        echo "EMBEDDED-HW: Raspberry Pi 5 detected." > /dev/kmsg
        PLATFORM_DIR="/embedded/rpi5"
        ;;
    *bcm2711*|*raspberrypi,4*)
        echo "EMBEDDED-HW: Raspberry Pi 4 detected." > /dev/kmsg
        PLATFORM_DIR="/embedded/rpi4"
        ;;
    *sifive*)
        echo "EMBEDDED-HW: SiFive hardware platform detected." > /dev/kmsg
        PLATFORM_DIR="/embedded/sifive"
        ;;
    *)
        echo "EMBEDDED-HW WARN: Unknown hardware platform ($COMPAT)." > /dev/kmsg
        PLATFORM_DIR=""
        ;;
esac

# 3. Dynamically fire off all scripts inside the designated hardware folder
if [ -n "$PLATFORM_DIR" ] && [ -d "$PLATFORM_DIR" ]; then
    for script in "$PLATFORM_DIR"/*.sh; do
        if [ -x "$script" ]; then
            echo "EMBEDDED-HW: Running $script..." > /dev/kmsg
            "$script"
        fi
    done
fi


#!/sh
# Early-stage initialization for a Waveshare or Sixfab 5G Cellular HAT on RPi5
# (Uses new RP1 driver structure on Raspberry Pi 5)

echo "EMBEDDED-HW: Power-cycling 5G Cellular HAT..." > /dev/kmsg

# Example: If your 5G HAT reset pin is tied to GPIO 6
# Verify if gpioset exists in the ramdisk to drive physical rails
if command -v gpioset >/dev/null; then
    # Pull Reset Low to clear state
    gpioset gpiochip4 6=0 2>/dev/null
    sleep 1
    # Pull Reset High to let the 5G modem boot clean in parallel with Linux
    gpioset gpiochip4 6=1 2>/dev/null
    echo "EMBEDDED-HW SUCCESS: 5G HAT hardware reset triggered via GPIO" > /dev/kmsg
else
    # Fallback to pure sysfs export if gpiod tools are missing
    if [ -d /sys/class/gpio ]; then
        echo 6 > /sys/class/gpio/export 2>/dev/null
        echo out > /sys/class/gpio/gpio6/direction 2>/dev/null
        echo 0 > /sys/class/gpio/gpio6/value
        sleep 1
        echo 1 > /sys/class/gpio/gpio6/value
    fi
fi

#!/sh
# Set up status LEDs or failsafe signals early

# Flash an external status line or buzzer attached to GPIO 26 to signal initrd is healthy
if command -v gpioset >/dev/null; then
    gpioset gpiochip4 26=1
fi

**Comment by devkontrol** [comment]
CC @jwrdegoede

**Comment by jwrdegoede** [comment]
ATM the biggest problem when it comes to things like RPi hats / Beagle Bone Capes / ... is that despite various attempts over the years adding a kernel API to load the device-tree overlays for the addon-boards from userspace has never been accepted into the mainline kernel.

So the DTBOs always need to be loaded + applied through some custom bootloader mechanism before the kernel starts.

E.g. this example code given above:

```
# Example: Copying to a runtime location inside the ramdisk
cp /boot/dtbs/sifive-custom.dtb /sys/firmware/fdt 2>/dev/null || true
```

Only works with out of tree patches. IMHO adding support to dracut upstream for a userspace API which has not been accepted upstream is not a good idea.

So taken DTBO loading out of the picture since there is no in tree solution for that, I don't think there is much left for dracut to do here.

The only case where dracut could possible be involved is when booting requires access to the addon board during boot. E.g. booting from storage on an addon card. But dracut should automatically pickup any standard (modules, firmware, fs) dependencies for the rootfs independent of it being on an addon board.
